doc/myrescue.1
.\" $Header: /cvsroot/myrescue/doc/myrescue.1,v 1.12 2008/06/01 23:42:55 kristofk Exp $
.TH MYRESCUE "8" "June 2008" "myrescue 0.9.6" "User Commands"
.SH NAME
myrescue \- Harddisc Rescue
.SH SYNOPSIS
.B myrescue 
.RB [ -b
.IR block-size ]
.RB [ -B
.IR bitmap-file ]
.RB [ -A ]
.RB [ -S ]
.RB [ -r
.IR retry-count ]
.RB [ -f
.IR skip-failed ]
.RB [ -s
.IR start-block ]
.RB [ -e
.IR end-block ]
.RB [ -R ]
.RB [ -G
.IR good-range ]
.RB [ -F
.IR failed-range ]
.RB [ -T ]
.RB [ -J
.IR jump-after-blocks ]
.I input-file
.I output-file
.SH DESCRIPTION
myrescue is a program to rescue the still-readable data from a
damaged harddisk. It is similiar in purpose to dd_rescue, but it
can be run in multiple passes and has options to avoid damaged areas
to first handle the not yet damaged part of the disk.
.PP
The program tries to copy the device blockwise to a file and
keeps a table ("block bitmap") noting whether a block has been
successfully copied, not yet handled or has had errors. This block
bitmap is used in successive passes to only read the not yet
rescued blocks.
.PP
The program has a special skip mode to handle read errors. Usually
harddisk surface defects cover more than just one block and continuous
reading in defect areas can damage the surface, the heads and (by
permanent recalibration) the drive mechanics. If this happens, the
chances of rescuing the remaining undamaged data drop dramatically. So
in skip mode, myrescue tries to get out of damaged areas quickly by
exponentially increasing the stepsize. The skipped blocks are marked
as unhandled in the block bitmap and can be retried later.
.PP
As another alternative, the program can jump around on the disc,
trying random blocks to first get an overview of the damage. In case
of regular defects (e.g. an entire failed head) this can be used to
determine the physical disc structure and to avoid these regions in
the first run.
.PP
Also there are options to avoid getting close to already recognized
defects or stay in the proximity of good regions.
.PP
Finally, the program has an option to multiply try to read a
block before considering it damaged.
.SH NOTE
.B This tools is no replacement for a professional data recovery service!
If you do have the latter option, don't even think of using 
.BR myrescue ,
as it may further damage your disk. This tool is
provided only for the case that you are absolutely desperate and
definitely cannot afford a professional data recovery and know what
you are doing. 
.PP
Data recovery - whether professional or DIY - is always careful
detective work. Damaged drives deteriorate rapidly and behave
nondeterministically, you may not get a second chance once you made a
mistake. So you have to be fully aware of what you're doing and
understand precisely what's happening. If you are unsure it's probably
best to stop right now and ask a linux guru for assistance.
.PP
In any case do not expect too much. While complete restores have been
witnessed, you should not take them for granted. A better attitude is
to consider your data lost and be glad for any survivors that turn up.
.PP
The usual GPL disclaimer applies. Especially the NON-WARRANTY OF
FITNESS FOR A PARTICULAR PURPOSE. Don't blame (or sue) me if it
fails to recover or further damages your data.
.PP
And a final word you probably don't want to hear in this situation:
For the future consider a routinely backup to avoid a "next time".
.SH OPTIONS
.TP
.BI -b " block-size"
The size of the blocks (in bytes). Set this to your
harddiscs error detection/correction unit size. Usually
this is 4096, which happens to be the default.
.TP
.BI -B " bitmap-file"
The file containing the status table of all blocks. Nice (or
frightening...) to view with hexdump. 01 means OK; 02 means that the
block was OK, but took a long time to read; 00 means not yet done;
negative values mean the number of failed read attempts. If not
given, defaults to
.IB output-file .bitmap
.TP
.B -A
Abort when encountering errors.
.TP
.B -S
Activate skip mode: When encountering errors increase the stepsize
exponentially until a readable block is found.
.TP
.BI -f " skip-failed"
Skip blocks that have already had 
.I skip-failed
failures. Useful to avoid scratching the same block over and over again.
.TP
.BI -r " retry-count"
The number of times to read a block before it is declared
bad for this run. (You can still retry it on the next run.) Default: 1
.TP
.BI -s " start-block"
The number of the block to start with. Default: 0
.TP
.BI -e " end-block"
The number of the block, where reading stops (not included!).
Default: size of 
.I input-file
divided by 
.IR block-size . 
.TP
.B -R
Reverse reading direction, i.e. from
.I end-block
(excluded) to
.I start-block
.TP
.BI -G " good-range"
Only try to read blocks within
.I good-range
blocks from an already successfully read block.
.TP
.BI -F " failed-range"
Extends -f to also skip any block within
.I failed-range
blocks of a block to be skipped as specified by -f.
.TP
.B -T
Also avoid blocks, that were successfully read, but took an unusually
long time to read. In case of growing defects, these blocks can
mark the boundaries of defective regions. This options makes -A, -S
and -F treat such blocks as defective.
.TP
.BI -J " jump-after-blocks"
Randomly jump across the disc after reading
.I jump-after-blocks
blocks in both directions. This might be useful to scan discs with scattered defects.
In jump mode -S causes myrescue to jump to a new block upon the first 
failed sector or upon hitting a sector to be skipped as specified by 
-f, -G or -F.
.TP
.BR -h ", " -?
Display usage information.
.SH RECOMMENDED PROCEDURE
NOTE: In data recovery every case needs special consideration and a
specific approach - mindlessly running procedures is a bad idea. So
consider the following an example, keep a close eye on the process and
be prepared to stop and reconsider if anything unusual happens. And
once again: If you are feeling unsure, stop and ask someone
experienced for assistance.
.IP \(bu
Make sure you have sufficient disk space for twice the whole
partition (complete partition size, not just the used amount of data)
plus some space for the block bitmap (1 byte per block). Of course
this should NOT be on the damaged disc.
.IP \(bu
Determine the hardware block size (CRC/ECC unit) of your
harddisk. This may be found out from hdparm, some entries in
/proc/ide/hd? or on the web. I have not yet checked whether
this is possible with an ioctl. If you have, please let me
know.
.IP \(bu
Start a skip mode run with one retry per block to first copy
the undamaged area.
.IP \(bu
Start a normal run with one retry per block to copy the
remaining skipped blocks. You may try to use -f 1 to skip the
damaged blocks from the first run.
.IP \(bu
Repeat until the number of errors seems to have converged.
Try waiting a couple of hours between the retries.
.IP \(bu
Repeat this with higher retry counts and wait for convergence.
.IP \(bu
.B Make a copy of the rescued data
and run fsck on it. The copy is important! Running fsck is risky if
blocks containing filesystem structures (superblocks, directories,
inodes, journal, ...) are missing. Sometimes the well-meant attempt to
correct the filesystem can damage the image file, so always work on a
copy.
.IP \(bu
Mount the filesystem (if copied to a file: via loopback) and check
your data. If directory information has been destroyed, fsck moves
unidentifiable file fragments to lost+found, so you should also check
this location.
.PP
It may help to try reading non-defect areas in between to allow the
drive to recalibrate.
.PP
The developers are glad to hear about your experiences. Please post
them to the
.B Experiences
forum on the Sourceforge Project page. Thank you!
.SH KNOWN BUGS
The handling of the 
.I bitmap-file 
currently relies on the filesystem semantics, that when
.BR lseek (2)
ing beyond the end of file and then writing, the space in between is
filled with zero-bytes.
.PP
The block bitmap maxes out after 127 failed read attempts.
.SH AUTHORS
Kristof Koehler <kristofk@users.sourceforge.net>,
Peter Schlaile <schlaile@users.sourceforge.net>
.SH SEE ALSO
.BR dd (1),
.BR dd_rescue "(no manpage?)"
.PP
http://www.google.de/search?q=data+recovery
.PP
http://myrescue.sourceforge.net/


src/myrescue.c
/*
    myrescue Harddisc Rescue Tool
    Copyright (C) 2002  Kristof Koehler (kristofk at users.sourceforge.net)
                        Peter Schlaile  (schlaile at users.sourceforge.net)

    This program is free software; you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation; either version 2 of the License, or
    (at your option) any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with this program; if not, write to the Free Software
    Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
*/

#define __USE_LARGEFILE64 1
#define _LARGEFILE_SOURCE 1
#define _LARGEFILE64_SOURCE 1

#include <stdio.h>
#include <stdlib.h>
#include <sys/types.h>
#include <unistd.h>
#include <time.h>
#include <errno.h>
#include <sys/types.h>
#include <sys/stat.h>
#include <fcntl.h>
#include <getopt.h>
#include <string.h>
#include <signal.h>

#include "permute.h"

#define LONG_TIME 3

int sig_abort = 0;

void handle_sig ( int dummy )
{
	sig_abort = 1;
}

long long filesize ( int fd )
{
	long long rval = lseek64(fd, 0, SEEK_END) ;
	if (rval < 0) {
		perror("filesize");
		exit(-1);
	}
	return rval;
}

int peek_map(int bitmap_fd, int block)
{
	char c = 0;
	if (lseek64(bitmap_fd, block, SEEK_SET) < 0) {
		perror("peek_map lseek64");
		exit(-1);
	}
	if (read(bitmap_fd, &c, 1) < 0) {
		perror("peek_map read");
		exit(-1);
	}
	return c;
}

void poke_map(int bitmap_fd, int block, char val)
{
	if (lseek64(bitmap_fd, block, SEEK_SET) < 0) {
		perror("poke_map lseek64");
		exit(-1);
	}
	if (write(bitmap_fd, &val, 1) != 1) {
		perror("poke_map write");
		exit(-1);
	}
}

int copy_block( int src_fd, int dst_fd, 
		long block_num, int block_size,
		unsigned char * buffer )
{
	long long filepos ;
	ssize_t src_count ;
	ssize_t dst_count ;

	filepos = block_num;
	filepos *= block_size;

	if (lseek64(src_fd, filepos, SEEK_SET) < 0) {
		perror("lseek64 src_fd");
		return errno;
	}

	if (lseek64(dst_fd, filepos, SEEK_SET) < 0) {
		perror("lseek64 dst_fd");
		return errno;
	}

	src_count = read(src_fd, buffer, block_size);
	if (src_count != block_size) {
		if (src_count == -1) {
			perror("src read failed");
			return errno;
		} else {
			fprintf(stderr,"short read: %d of %d\n",
				src_count, block_size);
			return -1;
		}
	}
		
	dst_count = write(dst_fd, buffer, block_size);
	if (dst_count != block_size) {
		if (dst_count == -1) {
			perror("dst write failed");
			return errno;
		} else {
			fprintf(stderr,"short write: %d of %d\n",
				dst_count, block_size);
			return -1;
		}
	}
	return 0;
}

int try_block ( int src_fd, int dst_fd, 
	       long block_num, int block_size, int retry_count,
		unsigned char * buffer )
{
	int r ;
	for ( r = 0 ; r < retry_count && ! sig_abort ; r++ ) {
		if ( copy_block ( src_fd, dst_fd, 
				  block_num, block_size,
				  buffer ) == 0 )
			return 1 ;
	}
	return 0 ;
}

int check_block ( int bitmap_fd, long block_num, 
		  int good_range, int failed_range,
		  int skip_fail, int long_as_bad,
		  long start, long end )
{
	int range = (good_range > failed_range) 
		? good_range : failed_range;
	int found_good = 0;
	long b;
	for ( b = block_num - range; b <= block_num + range; b++ ) {
		if ( (b < start) || (b >= end) )
			continue;
		char st = peek_map(bitmap_fd,b);
		if ( (failed_range > 0) &&
		     (abs(block_num-b) <= failed_range) &&
		     ((-st >= skip_fail) || (long_as_bad && st == 2)) )
			return 0;
		if ( (good_range > 0) &&
		     (abs(block_num-b) <= good_range) &&
		     ((st > 0) && (!long_as_bad || st != 2)) )
			found_good = 1;
	}
	return good_range > 0 ? found_good : 1;
}

void print_status ( long block, long start_block, long end_block, 
		    long ok_count, long bad_count, long done_count )
{
	fprintf ( stderr, 
		  "\rblock %09ld (%09ld-%09ld, %.2f%%)   "
		  "ok %09ld   bad %09ld   ",
		  block, start_block, end_block,
		  (100.0*done_count)/(end_block-start_block),
		  ok_count, bad_count ) ;
}

void do_copy ( int src_fd, int dst_fd, int bitmap_fd,
	       int block_size, long start_block, long end_block,
	       int retry_count, int abort_error, int skip, 
	       int skip_fail, int reverse,
	       int good_range, int failed_range, int long_as_bad,
	       unsigned char * buffer )
{
	long block_step = 1;
	long block ;
	long ok_count = 0 ;
	long bad_count = 0 ;
	char block_state ;
	int  forward = !reverse;
	time_t before, after;
	int long_time;

	for ( block = forward ? start_block : (end_block-1) ; 
	      (forward ? (block < end_block) : (block >= start_block)) && ! sig_abort ; 
	      block += forward ? block_step : -block_step ) {
		block_state = peek_map ( bitmap_fd, block ) ;
		if ( (block_state <= 0) &&
		     ( (skip_fail == 0) || (-block_state < skip_fail) ) &&
		     check_block ( bitmap_fd, block,
				   good_range, failed_range, 
				   skip_fail, long_as_bad,
				   start_block, end_block ) ) {
			print_status ( block, start_block, end_block, 
				       ok_count, bad_count, 
				       forward ? block-start_block : end_block-1-block );
			time(&before);
			if ( try_block ( src_fd, dst_fd, 
					 block, block_size, retry_count,
					 buffer ) ) {
				time(&after);
				long_time = (after - before >= LONG_TIME);
				++ok_count ;
				poke_map(bitmap_fd, block, long_time ? 2 : 1);
				if ( long_time && long_as_bad ) {
					if (abort_error)
						break;
					if (skip)
						block_step *= 2;
				} else {
					block_step = 1;
				}
			} else {
				++bad_count;
				if ( block_state > -127 )
					block_state -= 1;
				poke_map(bitmap_fd, block, block_state);
				if (abort_error)
					break;
				if (skip)
					block_step *= 2;
			}
		} else {
			if ( block % 1000 == 0 ) {
				print_status ( block, start_block, end_block, 
					       ok_count, bad_count, 
					       forward ? block-start_block : end_block-1-block );
			}
			block_step = 1 ;
		}
		
	}
	if ( sig_abort )
		fprintf ( stderr, "\nABORTED    " );
	else
		print_status ( forward ? end_block : start_block, start_block, end_block, 
		       ok_count, bad_count, end_block-start_block ) ;
	fprintf ( stderr, "\n" ) ;
}

int do_jump_run ( int src_fd, int dst_fd, int bitmap_fd,
		  int block_size, long start_block, long end_block,
		  int retry_count, int abort_error, int skip, 
		  int skip_fail, int jump,
		  int good_range, int failed_range, int long_as_bad,
		  long block, long orig_block,
		  int jump_count, int jump_step,
		  long *ok_count, long *bad_count,
		  unsigned char * buffer ) 
{
	char block_state ;
	time_t before, after;
	int long_time;

	for ( ; jump_count-- > 0 && !sig_abort ; block += jump_step ) {
		if ( block >= end_block )
			break;
		if ( block < start_block )
			break;
		
		block_state = peek_map ( bitmap_fd, block ) ;
		if ( block_state > 0 )
			continue;
		
		if ( ((skip_fail > 0) && (-block_state >= skip_fail)) ||
		     (!check_block ( bitmap_fd, block,
				     good_range, failed_range, 
				     skip_fail, long_as_bad,
				     start_block, end_block )) ) {
			if (skip || abort_error)
				return 0;
			else
				continue;
		}

		print_status ( block, start_block, end_block, 
			       *ok_count, *bad_count, orig_block-start_block ) ;
		
		time(&before);
		if ( try_block ( src_fd, dst_fd, 
				 block, block_size, retry_count,
				 buffer ) ) {
			time(&after);
			long_time = (after - before >= LONG_TIME);
			++(*ok_count);
			poke_map(bitmap_fd, block, long_time ? 2 : 1);
			if ( long_as_bad && long_time ) {
				if (skip || abort_error)
					return 0;
			}
		} else {
			++(*bad_count);
			if ( block_state > -127 )
				block_state -= 1;
			poke_map(bitmap_fd, block, block_state);
			if (skip || abort_error)
				return 0;
		}
	}
	return 1;
}

void do_jump ( int src_fd, int dst_fd, int bitmap_fd,
	       int block_size, long start_block, long end_block,
	       int retry_count, int abort_error, int skip, 
	       int skip_fail, int jump,
	       int good_range, int failed_range, int long_as_bad,
	       unsigned char * buffer )
{
	long orig_block, block ;
	long ok_count = 0 ;
	long bad_count = 0 ;
	permute_info *pi;

	srandom(getpid() ^ time(NULL));

	pi = permute_init(997,16);
	
	for ( orig_block = start_block; orig_block < end_block && !sig_abort; orig_block++ ) {

		block = permute ( orig_block-start_block,
				  end_block-start_block,
				  pi ) + start_block;

		if ( orig_block % 1000 == 0 )
			print_status ( block, start_block, end_block, 
				       ok_count, bad_count, 
				       orig_block-start_block );
		
		if ( ! do_jump_run(src_fd, dst_fd, bitmap_fd,
				   block_size, start_block, end_block,
				   retry_count, abort_error, skip, 
				   skip_fail, jump,
				   good_range, failed_range, long_as_bad,
				   block, orig_block,
				   jump, +1,
				   &ok_count, &bad_count,
				   buffer) )
			if ( abort_error )
				break;

		if ( ! do_jump_run(src_fd, dst_fd, bitmap_fd,
				   block_size, start_block, end_block,
				   retry_count, abort_error, skip, 
				   skip_fail, jump,
				   good_range, failed_range, long_as_bad,
				   block-1, orig_block,
				   jump-1, -1,
				   &ok_count, &bad_count,
				   buffer) )
			if ( abort_error )
				break;
	}
	permute_free(pi);

	if ( sig_abort )
		fprintf ( stderr, "\nABORTED    " );
	else
		print_status ( end_block, start_block, end_block, 
			       ok_count, bad_count, end_block-start_block ) ;
	fprintf(stderr,"\n");
}

const char * usage = 
"myrescue [<options>] <input-file> <output-file>\n"
"options:\n"
"-b <block-size>   block size in bytes, default: 4096\n"
"-B <bitmap-file>  bitmap-file, default: <output-file>.bitmap\n"
"-A                abort on error\n"
"-S                skip errors (exponential-step)\n"
"-f <number>       skip blocks with <number> or more failures\n"
"-r <retry-count>  try up to <retry-count> reads per block, default: 1\n"
"-s <start-block>  start block number, default: 0\n"
"-e <end-block>    end block number (excl.), default: size of <input-file>\n"
"-G <range>        only read <range> blocks around good ones\n"
"-F <range>        skip <range> blocks around failed ones\n"
"-J <number>       randomly jump after reading a few sectors\n"
"-T                make -A, -S and -F avoid blocks that took long to read\n"
"-R                reverse copy direction\n"
"-h, -?            usage information\n" ;

int main(int argc, char** argv)
{
	char *src_name ;
	char *dst_name ;
	char *bitmap_name = NULL ;
	char bitmap_suffix[] = ".bitmap" ;
	
	int block_size    = 4096 ;
	int abort_error   = 0 ;
	int skip          = 0 ;
	int skip_fail     = 0 ;
	int retry_count   = 1 ;
	long start_block  = 0 ;
	long end_block    = -1 ;
	int reverse       = 0 ;
	int jump          = 0 ;
	int good_range    = 0 ;
	int failed_range  = 0 ;
	int long_as_bad    = 0 ;

	long long block_count ;

	int dst_fd ;
	int src_fd ;
	int bitmap_fd ;

	unsigned char* buffer ;

        int optc ;

	/* options */

        while ( (optc = getopt ( argc, argv, "b:B:ASf:r:s:e:J:G:F:TRh?" ) ) != -1 ) {
		switch ( optc ) {
		case 'b' :
			block_size = atol(optarg);
			if (block_size <= 0) {
				fprintf(stderr, "invalid block-size: %s\n", 
					optarg);
				exit(-1);
			}
			break ;
		case 'B' :
			bitmap_name = optarg;
			break ;
		case 'A' :
			abort_error = 1 ;
			break ;
		case 'S' :
			skip = 1 ;
			break ;
		case 'f' :
			skip_fail = atol(optarg);
			if (skip_fail <= 0) {
				fprintf(stderr, "invalid skip-failed level: %s\n", 
					optarg);
				exit(-1);
			}
			break ;
		case 'r' :
			retry_count = atol(optarg);
			if (retry_count <= 0) {
				fprintf(stderr, "invalid retry-count: %s\n", 
					optarg);
				exit(-1);
			}
			break ;
		case 's' :
			start_block = atol(optarg);
			if (start_block <= 0) {
				fprintf(stderr, "invalid start_block: %s\n", 
					optarg);
				exit(-1);
			}
			break ;
		case 'e' :
			end_block = atol(optarg);
			if (end_block <= 0) {
				fprintf(stderr, "invalid end_block: %s\n", 
					optarg);
				exit(-1);
			}
			break ;
		case 'J' :
			jump = atol(optarg);
			if (jump <= 0) {
				fprintf(stderr, "invalid jump value: %s\n", 
					optarg);
				exit(-1);
			}
			break ;
		case 'G' :
			good_range = atol(optarg);
			if (good_range <= 0) {
				fprintf(stderr, "invalid good range value: %s\n",
					optarg);
				exit(-1);
			}
			break ;
		case 'F' :
			failed_range = atol(optarg);
			if (failed_range <= 0) {
				fprintf(stderr, "invalid failed range value: %s\n",
					optarg);
				exit(-1);
			}
			break ;
		case 'T' :
			long_as_bad = 1 ;
			break ;
		case 'R' :
			reverse = 1 ;
			break ;
                default :
			fprintf ( stderr, "%s", usage ) ;
			exit(-1) ;
		}
	}
        if (optind != argc - 2) {
		fprintf ( stderr, "%s", usage ) ;
		exit(-1) ;
	}

	/* buffer */
	buffer = malloc ( block_size ) ;
	if ( buffer == NULL ) {
		fprintf ( stderr, "malloc (%d) failed\n", block_size ) ;
		exit(-1) ;
	}

	/* filenames */

	src_name = argv[optind] ;
	dst_name = argv[optind+1] ;

	if ( bitmap_name == NULL ) {
		bitmap_name = malloc ( strlen(dst_name) + 
				       strlen(bitmap_suffix) + 1 ) ;
		if ( bitmap_name == NULL ) {
			fprintf ( stderr, "malloc failed\n" ) ;
			exit(-1) ;
		}
		strcpy ( bitmap_name, dst_name ) ;
		strcat ( bitmap_name, bitmap_suffix ) ;
	}

	/* open files */

	src_fd = open64(src_name, O_RDONLY);
	if ( src_fd < 0 ) {
		perror ( "source open failed" ) ;
		exit(-1) ;
	}

	dst_fd = open64(dst_name, O_RDWR | O_CREAT, 0600);
	if ( dst_fd < 0 ) {
		perror ( "destination open failed" ) ;
		exit(-1) ;
	}

	bitmap_fd = open64(bitmap_name, O_RDWR | O_CREAT, 0600);
	if ( bitmap_fd < 0 ) {
		perror ( "bitmap open failed" ) ;
		exit(-1) ;
	}

	/* maximum block */
	block_count = filesize(src_fd) ;
	block_count /= block_size ;

	if ( end_block == -1 ) {
		end_block = block_count ;
	} 
#ifdef CHECK_END_BLOCK
	if ( end_block > block_count ) {
		fprintf ( stderr, 
			  "end_block(%ld) > block_count(%Ld)\n"
			  "end_block clipped\n",
			  end_block, block_count ) ;
		end_block = block_count ;
	}
#endif
	if ( start_block >= end_block ) {
		fprintf ( stderr, 
			  "start_block(%ld) >= end_block(%ld)\n",
			  start_block, end_block ) ;
		exit(-1) ;
	}

	/* start the real job */
	signal(SIGINT, handle_sig);
	if ( jump == 0 )
		do_copy ( src_fd, dst_fd, bitmap_fd,
			  block_size, start_block, end_block,
			  retry_count, abort_error, skip, 
			  skip_fail, reverse,
			  good_range, failed_range, long_as_bad,
			  buffer ) ;
	else
		do_jump ( src_fd, dst_fd, bitmap_fd,
			  block_size, start_block, end_block,
			  retry_count, abort_error, skip, 
			  skip_fail, jump,
			  good_range, failed_range, long_as_bad,
			  buffer );
	return 0 ;
}
